main/src/library/RedBlackTree.flix

```diff
@@ -474,11 +474,66 @@ namespace RedBlackTree {
         case _ => 0
     }
 
+    // Note: The use of a multiplier. Using 1 thread per logic core is too little.
+    def multiplier(): Int32 = 4
+
     ///
     /// Returns a RedBlackTree with mappings `k -> f(v)` for every `k -> v` in `tree`.
     ///
-    pub def mapWithKey(f: (k, v1) -> v2 & ef, tree: RedBlackTree[k, v1]): RedBlackTree[k, v2] & ef = match tree {
-        case Node(color, left, k, v, right) => Node(color, mapWithKey(f, left), k, f(k, v), mapWithKey(f, right))
-        case _                              => Leaf
+    pub def mapWithKey(f: (k, v1) -> v2 & ef, t: RedBlackTree[k, v1]): RedBlackTree[k, v2] & ef =
+        let cpus = multiplier() * Environment.getVirtualProcessors();
+        let f1 = (r) -> f(r.k, r.v);
+        reifyEff(f1) {
+            case Pure(g) =>
+                let g1 = (k, v) -> g({k = k, v = v});
+                if (RedBlackTree.blackHeight(t) >= 4)
+                    parMap(cpus - 1, g1, t) as & Pure
+                else
+                    seqMap(g1, t)
+            case _  => seqMap(f, t)
+        }
+
+    pub def parMap(n: Int32, f: (k, v1) -> v2, t: RedBlackTree[k, v1]): RedBlackTree[k, v2] & Impure = {
+        if (n <= 1)
+            parMapSeq(f, t)
+        else
+            match t {
+                case Leaf                   => Leaf
+                case DoubleBlackLeaf        => DoubleBlackLeaf
+                case Node(c, a, k, v, b)    =>
+                    let chanL = chan RedBlackTree[k, v2] 1;
+                    let chanR = chan RedBlackTree[k, v2] 1;
+                    spawn (chanL <- parMap((n - 2) / 2, f, a));
+                    spawn (chanR <- parMap((n - 2) / 2, f, b));
+                    let v1 = f(k, v);
+                    let a1 = <- chanL;
+                    let b1 = <- chanR;
+                    Node(c, a1, k, v1, b1)
+            }
     }
+
+    // Note that while this is identical to seqMap it must still be its own function
+    // because the parallel path must not be reuse the seq. path.
+    def parMapSeq(f: (k, v1) -> v2 & ef, t: RedBlackTree[k, v1]): RedBlackTree[k, v2] & ef = match t {
```

Comment:
```
///
/// Maps `f` over the tree `t` sequentially from left to right.
///
/// Note that this is equivalent to `seqMap` but for performance reasons -- to avoid
/// megamorphic calls -- we use a copy here.
/// 
```

Reply:
I would also change `parMapSeq` to require a pure function.

Reply:
Non-public btw?